### PR TITLE
Implement updateSession & isUpdate parameter

### LINKED ISSUE #596: getSession call does not trigger session object update until window loses focus
I'm currently trying to update some page components when the session changes after an `onClick`-triggered `Promise` is fulfilled. Here the function `DeleteGroup` returns a `Promise`. I print the value of the session before the `Promise` is returned, and then print it again after it is fulfilled by calling `getSession()`.

```jsx
<Button
  onClick={() => {
    console.log(session);
    DeleteGroup(value.id).then(async () => console.log(await getSession()));
  }}
>
```

According to the console, everything I've logged looks right. The initial `console.log(session)` output shows `groups` as an `Array` of length 2, while the second `console.log(await getSession())` output shows it as length 1.

Screen Shot 2020-08-20 at 6 02 35 PM

However, until the window loses focus, my UI continues to show 2 groups despite the console output clearly showing the result of the `getSession` call having only 1 group. My UI is immediately synced once the window loses focus, so I do not believe there is anything wrong with the way I've written my UI components. If that is the case, why wouldn't my `getSession()` call immediately update the UI accordingly?

Apologies if this is a repeat; I find session updating to be the trickiest part of Next-Auth.

### LINKED ISSUE #2129: Session is not renewed on client side right after SSR state changes
## Description 🐜

The problem is described here: https://github.com/nextauthjs/next-auth/discussions/2123#discussion-3395145

Seems like when changing the session at server side, the client will still use the old session states. 
This leads to false informations at the client side. 

## How to reproduce ☕️

https://codesandbox.io/s/4gokt?file=/pages/_app.tsx

Changes is at `[...nextauth]`, `_app` and `client`.
_app has getinitialprops, which uses the getSession method. 
It will at the server side and client side use the same state, but i expect the client side to use new the new state. 

## Environment 🖥

```
npx: installed 1 in 1.676s

  System:
    OS: macOS 11.4
    CPU: (16) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Memory: 854.63 MB / 16.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 14.16.0 - ~/.nvm/versions/node/v14.16.0/bin/node
    Yarn: 1.22.10 - /usr/local/bin/yarn
    npm: 6.14.11 - ~/.nvm/versions/node/v14.16.0/bin/npm
    Watchman: 4.9.0 - /usr/local/bin/watchman
  Browsers:
    Chrome: 91.0.4472.77
    Safari: 14.1.1
```

### LINKED ISSUE #2208: useSession/Provider does not update when a new session is created until page reload.
## Description 🐜

When a page is not reloaded after login, the useSession/Provider is never updated, even after calling getSession.

## How to reproduce ☕️

1. Here is a [reproduction ](https://github.com/PaulPCIO/next-auth-session) of the issue.
2. I have set up a modal based login using a security code, for this example, I have [hardcoded ](https://github.com/PaulPCIO/next-auth-session/blob/434b417d24e3ae1d4b1190db2a97fb6da169c296/pages/api/auth/%5B...nextauth%5D.js#L14) the security code to `TEST`
3. When you log in successfully, the [callpack API](https://github.com/PaulPCIO/next-auth-session/blob/434b417d24e3ae1d4b1190db2a97fb6da169c296/components/login-view.js#L95-L97) triggers an additional [getSession](https://github.com/PaulPCIO/next-auth-session/blob/434b417d24e3ae1d4b1190db2a97fb6da169c296/components/login-view.js#L99).
4. At this stage the cookies are created and the session should be there, but the provider does not update until you refresh the page.

## Screenshots / Logs 📽

example

## Environment 🖥

```
  System:
    OS: Linux 5.4 Ubuntu 20.04.2 LTS (Focal Fossa)
    CPU: (8) x64 Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
    Memory: 11.23 GB / 24.97 GB
    Container: Yes
    Shell: 5.0.17 - /bin/bash
  Binaries:
    Node: 14.16.1 - ~/.nvm/versions/node/v14.16.1/bin/node
    Yarn: 1.22.10 - ~/.nvm/versions/node/v14.16.1/bin/yarn
    npm: 6.14.12 - ~/.nvm/versions/node/v14.16.1/bin/npm
```

### PR DESCRIPTION
**Pull request by christoph-pflueger**
## ☕️ Reasoning

What changes are being made? What feature/bug is being fixed here?

Implements `updateSession` to be able to update the `SessionProvider`'s state without having to reload the page. The update event is propagated to the `jwt` and `session` callbacks via an `isUpdate` parameter. This helps to avoid having to check if the token and session is up to date on every request.

As there are currently no tests implemented for callbacks, I skipped that step.

## 🧢 Checklist

- [x] Documentation
- [ ] Tests
- [ ] Ready to be merged

## 🎫 Affected issues

Please [scout and link issues](https://github.com/nextauthjs/next-auth/issues) that might be solved by this PR.

Fixes: #596 
Fixes: #3941
Fixes: #2208
Might fix: #2129

## 📌 Resources

- [Contributing guidelines](./CONTRIBUTING.md)
- [Code of conduct](./CODE_OF_CONDUCT.md)
- [Contributing to Open Source](https://kcd.im/pull-request)

### CODE REVIEW

**Comment by DitlevHB** [comment]
Any ETA for the release of this?

**Comment by qiqo** [comment]
Can I gently nudge this pull request? I hope if gets merged already as this is also a blocker in our application - I think this should be considered critical

**Comment by git-chrisp** [comment]
> Can I gently nudge this pull request? I hope if gets merged already as this is also a blocker in our application - I think this should be considered critical

The fact that this pull request hasn't received as much as a reply for almost three months even though it has been requested multiple times is mind-blowing to me. It's not like other pull requests are not being worked on.

**Comment by qiqo** [comment]
> > Can I gently nudge this pull request? I hope if gets merged already as this is also a blocker in our application - I think this should be considered critical
> 
> The fact that this pull request hasn't received as much as a reply for almost three months even though there appears to be genuine interest is mind-blowing to me. It's not like other pull requests are not being worked on either.

and the fix isn't even outrageously complex as well it's practically  just assistive code to extend session's functionality, it just needs a quick nod/verify from them. The inaction is like compelling users to fork and manually implement the fix.

**Comment by devrsi0n** [comment]
@balazsorban44 Could you help review this PR? It seems like a common feature should be built into next-auth. Thank you for your time.

**Comment by qiqo** [comment]
@balazsorban44 hoping for your time and consideration on this fixes. Thanks

**Review by ThangHuuVu — COMMENTED** [review]
Thanks for the PR @git-chrisp  and sorry for the delayed response, I reviewed your PR
It seems like the added method  `updateSession` is exactly the same as `getSession`, the only difference is in the callback `session` and `jwt`
Could you give an example case where the extra `isUpdate` param is helpful? 🤔

**Comment by git-chrisp** [comment]
> Thanks for the PR @git-chrisp and sorry for the delayed response, I reviewed your PR It seems like the added method `updateSession` is exactly the same as `getSession`, the only difference is in the callback `session` and `jwt` Could you give an example case where the extra `isUpdate` param is helpful? 🤔

The implemented `updateSession` doesn't call `getSession`. It calls the `__NEXTAUTH._getSession` which is encapsulated inside the `SessionProvider` and cannot be accessed at the moment.

Assume someone changes something in his profile settings that is stored in e.g. the token's custom claims. There is no way to update this change in the `SessionProvider` at the moment without reloading the entire page.

The `isUpdate` parameter helps in the backend to differentiate between normal `getSession`, `_getSession` and `updateSession` calls.

**Comment by shahryarjb** [comment]
Is there any deadline to fix this problem? For now the only way to fix this issue is reloading page and this way clean the state or the action user wants to do
Thank you.

**Comment by mohamedGamalAbuGalala** [comment]
Is any update on this pr soon, we are having the same problem authenticating with a custom PKCE provider, and when refreshing the token and have the new token back. the client keeps calling JWT with the old invalid token. and we use one-time refresh tokens so, it's pretty impossible to get around this setup without any compromise with the rules we already have.

**Comment by rustunooldu** [comment]
May I kindly ask for an update? I had this PR bookmarked months ago, and it's a little disheartening to see that there hasn't been any progress... It's a blocker for our application, and there seems to be a lot of interest from other people as well.

**Comment by raulmarindev** [comment]
I'm sorry if I'm mistaken but this related to this workaround? https://stackoverflow.com/a/70405437/9123724

**Comment by lemcii** [comment]
Currently using `window.refresh`/`window.location.pathname` since I cannot await session update after updating a users profile. The changes here would likely solve that.

**Comment by shahryarjb** [comment]
> window.location.reload()

It is not good, I need to update my auth 2 tokens {refresh token, access token} every 1 hour, but for now I am forced to reload my site: [it is my next auth API](https://github.com/mishka-group/mishka-cms-front/blob/master/pages/api/auth/%5B...nextauth%5D.js), and it is my client:

https://github.com/mishka-group/mishka-cms-front/blob/master/apps/mishka_user/helper/authHelper.ts#L121-L123

**Comment by rolanday** [comment]
Happy to send reviewers a case of beer to get this PR approved (and pulled). As others have noted, this is basic/needed functionality. Thanks!

File: docs/docs/configuration/callbacks.md [hunk 1]
```diff
@@ -22,10 +22,10 @@ You can specify a handler for any of the callbacks below.
     async redirect({ url, baseUrl }) {
       return baseUrl
     },
-    async session({ session, user, token }) {
+    async session({ session, user, token, isUpdate }) {
```
Comment (devrsi0n) [inline docs/docs/configuration/callbacks.md:25]: Could you use `isUpdated` instead? `isUpdate` has a grammar issue.

**Comment by kblizeck** [comment]
> 

I'd pitch in for a second case - this would be super beneficial!

**Comment by balazsorban44** [comment]
Closing in favor of #7056

**Pull request closed without merging**